lib/logmsg/nvtable.c

```diff
@@ -802,3 +802,60 @@ nv_table_clone(NVTable *self, gint additional_space)
 
   return new;
 }
+
+
+static gboolean
+_compact_foreach_entry(NVHandle handle, NVEntry *entry, NVIndexEntry *index_entry, gpointer user_data)
+{
+  gpointer *args = (gpointer *) user_data;
+  NVTable *old = (NVTable *) args[0];
+  NVTable *new = (NVTable *) args[1];
+  const gchar *value, *name;
+  gssize value_len, name_len;
+
+  /* unused entries are skipped */
+  if (entry->unset)
+    return FALSE;
+
+  if (entry->name_len)
+    {
+      /* non-builtin entries have their name stored in the origin NVTable, use that */
+      name = nv_entry_get_name(entry);
+      name_len = entry->name_len;
+    }
+  else
+    {
+      /* builtin entries don't have their name stored, but we won't store
+       * them either, so just set them to NULL/0 */
+      name = NULL;
+      name_len = 0;
+    }
+
+  if (!entry->indirect)
+    {
+      value = nv_table_resolve_direct(old, entry, &value_len);
+
+      gboolean value_successfully_added = nv_table_add_value(new, handle, name, name_len, value, value_len, NULL);
+      g_assert(value_successfully_added);
```

Comment:
This case is completely unexpected, we are preallocating the NVTable with a size that should be able to hold all name-value pairs (see that we are using the same size that was used by the original NVTable, which we should only use less of).  So if the add_value() call fails, it means that our preallocation in nv_table_alloc() is wrong.

This should be fixed instead of adding a reallocation loop here.

Reply:
I do not get how/why reallocation loop is mentioned here. The part before that is clearly answers my concern, which was not correct.

Reply:
nv_table_add_value() failures need to be handled by calling nv_table_realloc() in a loop, that's what I meant in that sentence. See an example here:

https://github.com/syslog-ng/syslog-ng/blob/a51919016c4e80d6d5027d25cb91bc41f63746a0/lib/logmsg/logmsg.c#L555